pkg/executables/cmk.go

```diff
@@ -0,0 +1,363 @@
+		return "", fmt.Errorf("template %s not found", template)
+	}
+	return templates[0].Name, nil
+}
+
+func (c *Cmk) SearchComputeOffering(ctx context.Context, domain string, zone string, account string, computeOffering string) (string, error) {
+	offerings, err := c.ListServiceOfferings(ctx, computeOffering)
+	if err != nil {
+		return "", fmt.Errorf("compute offering %s not found. error: %v", computeOffering, err)
+	} else if len(offerings) > 1 {
+		return "", fmt.Errorf("duplicate compute offering %s found", computeOffering)
+	} else if len(offerings) == 0 {
+		return "", fmt.Errorf("compute offering %s not found", computeOffering)
+	}
+
+	return offerings[0].Name, nil
+}
+
+func (c *Cmk) SearchDiskOffering(ctx context.Context, domain string, zone string, account string, diskOffering string) (string, error) {
+	if diskOffering == "" {
+		return diskOffering, nil
+	}
+	offerings, err := c.ListDiskOfferings(ctx, diskOffering)
+	if err != nil {
+		return "", fmt.Errorf("disk offering %s not found. error: %v", diskOffering, err)
+	} else if len(offerings) > 1 {
+		return "", fmt.Errorf("duplicate disk offering %s found", diskOffering)
+	} else if len(offerings) == 0 {
+		return "", fmt.Errorf("disk offering %s not found", diskOffering)
+	}
+	return offerings[0].Name, nil
+}
+
+func (c *Cmk) SearchAffinityGroups(ctx context.Context, domain string, zone string, account string, affinityGroupIds []string) error {
+	for _, affinityGroupId := range affinityGroupIds {
+		affinityGroup, err := c.ListAffinityGroupsById(ctx, affinityGroupId)
+		if err != nil {
+			return fmt.Errorf("affinity group %s not found. error: %v", affinityGroupId, err)
+		} else if len(affinityGroup) > 1 {
+			return fmt.Errorf("duplicate affinity group %s found", affinityGroupId)
+		} else if len(affinityGroup) == 0 {
+			return fmt.Errorf("affinity group %s not found", affinityGroupId)
+		}
+	}
+	return nil
+}
+
+// cmkExecConfig contains transient information for the execution of cmk commands
+// It must be cleaned after each execution to prevent side effects from past executions options
+type cmkExecConfig struct {
+	env        map[string]string
+	ConfigFile string
+}
+
+func NewCmk(executable Executable, writer filewriter.FileWriter) *Cmk {
+	return &Cmk{
+		writer:     writer,
+		executable: executable,
+		retrier:    retrier.NewWithMaxRetries(maxRetriesCmk, backOffPeriodCmk),
+	}
+}
+
+// ValidateCloudStackConnection Calls `cmk sync` to ensure that the endpoint and credentials + domain are valid
+func (c *Cmk) ValidateCloudStackConnection(ctx context.Context) error {
+	buffer, err := c.exec(ctx, "sync")
+	if err != nil {
+		return fmt.Errorf("error validating cloudstack connection for cmk config %s: %v", buffer.String(), err)
+	}
+	logger.MarkPass("Connected to CloudStack server")
+	return nil
+}
+
+// TODO: Add support for domain, account filtering
+func (c *Cmk) ListTemplates(ctx context.Context, template string) ([]types.CmkTemplate, error) {
```

Comment:
I find a bit confusing this method
If it's listing templates, why does it take a template as input?
I think I'm missing something :)

Reply:
The confusion is understandable. The method naming is based on how the cloudmonkey CLI itself is implemented. They don't expose get operations, only list with filters. I was trying to keep as much business logic as possible outside of these methods so they could be used easily by functions implementing business logic. If that's more confusing that valuable, I'll refactor to just have something like "ValidateTemplate" which makes sure there is exactly one template matching the criteria, otherwise return an error. And same for the other resource types.

Reply:
Oh I 100% support leaving the business logic out of this file. In fact out of `executables` completely (this is a mistake we have made in the past).

I think writing methods that follow `cmk` api is totally fine. Maybe we can just rename things a bit to avoid confusion among people who don't know a lot about the API (like me). For example, instead of `ListTemplates` we could go with `FilterTemplates`. And instead of `template` we could use `filterValue`.

That said, what is the reason the filter value can be name or ID? Doesn't the caller know what is the value they have when calling this method? We could just have different methods, like `FilterTemplatesByName` and `FilterTemplatesByID`.